### PR TITLE
enhancements and refinements

### PR DESCRIPTION
**Pull request by zayden-nexus**
Enhancements and Refinements

This PR introduces several enhancements and refinements to the `BridgeStandard`, `BridgeCross`, `BridgeEthereum` contracts, focusing on improving its functionality, clarity, and testability.

**1. Core Contract Functionality:**

***a. BridgeStandard Contract:***

* **Core Functionality:** Provides a standardized framework for cross-chain bridging, including bridge initiation, permit-based bridging, finalization, retry mechanisms, and fee management.
* **Logic Improvements:**
    * Zero address check for `rewardWallet` during initialization.
    * More descriptive revert messages for improved debugging.
    * Allows `bridgeTokenInfo` to be unset (zero address).
    * Clarified success conditions for `_permitCall` to handle various ERC20 permit implementations.
* **Test Code Improvements:** Expanded test coverage for core functionalities and edge cases, with improved test clarity.
* **Enhanced Documentation:** Comprehensive comments for public and internal functions, clarifying complex logic and using NatSpec for better documentation generation.

***b. BridgeCross Contract:***

* **Core Functionality:** Implements `BridgeStandard` for a specific cross-chain bridge, handling initiation and finalization of transfers, including minting/burning of wrapped tokens.  Uses `Create2` for deploying new wrapped tokens.
* **Logic Improvements:**
    * Handles native token transfers (e.g., ETH, BNB) and ERC20 token transfers.
    * Uses `Create2` for deterministic deployment of wrapped tokens via `addTokenDeploy`.
    * Implements `_initiateBridge` to burn tokens on the source chain or handle native token transfers and fees.
    * Implements `_finalizeBridge` to mint wrapped tokens on the destination chain or handle native token transfers. Includes comprehensive error handling and detailed revert reasons for minting failures.
* **Test Code Improvements**: Added test coverage for `Create2` deployment and cross-chain transfers with both native and wrapped tokens. Includes testing of edge cases and error handling.

***c. BridgeEthereum Contract:***

* **Core Functionality:**  Implements `BridgeStandard` for bridging from Ethereum, focusing on initiating transfers of the wrapped native token (`cross`, e.g., WETH) and finalizing transfers to Ethereum.
* **Logic Improvements:**
    * Implements `bridgeCross` and `bridgeToCross` for simplified bridge initiation on Ethereum, using the permit functionality.
    * Implements `_initiateBridge` to transfer and handle fees for the wrapped native token on Ethereum.
    * Implements `_finalizeBridge` to transfer tokens to the recipient on Ethereum. Includes error handling and revert reasons. Adjusts value based on the `_exrate` if the token is the wrapped native token.
* **Test Code Improvements:** Added test cases covering `bridgeCross`, `bridgeToCross`, and both successful and failing finalizations.  Tests cover various transfer amounts and edge cases.

These changes enhance the functionality, security, and maintainability of the bridging contracts.  The comprehensive test suites and improved documentation provide greater confidence in the contracts' correctness and facilitate easier integration, auditing, and understanding.

**2. Logic Improvements:**

* **Zero Address Check for `rewardWallet`:** Added `require` statement in initialization to ensure `rewardWallet` is not set to the zero address.
* **More descriptive `require` error messages:** Improved the clarity of revert messages to aid in debugging and understanding the cause of failures.
* **`bridgeTokenInfo` Zero Address Allowed**: The contract now allows the BridgeTokenInfo to be unset.
* **Clarified Success Conditions for `_permitCall`**: Added checks for various ERC20 permit implementations which may or may not return data.

**3. Test Code Improvements:**

* **Expanded Test Coverage:**  Increased test coverage for core functionalities, including:
    - Bridge initiation with and without permit.
    - Finalization success and failure scenarios.
    - Retry mechanism.
    - Fee calculations and distribution.
    - Edge cases, such as zero values and invalid signatures.
* **Improved Test Clarity:**  Used more descriptive test case names and added comments for better readability and understanding.

**4. Enhanced Documentation (Comments):**

* **Comprehensive Comments for Public and Internal Functions:** Added detailed explanations of function parameters, return values, and potential revert conditions.
* **Clarified Complex Logic:** Added comments to explain intricate parts of the code, such as the permit call and assembly sections.
* **NatSpec Comments:** Added or improved NatSpec comments to generate better documentation automatically.

### CODE REVIEW

File: src/BridgeCross.sol [hunk 1]
```diff
@@ -17,61 +23,101 @@ contract BridgeCross is BridgeStandard {
...
+     */
     function _initiateBridge(IERC20 token, address from, uint value, uint fee) internal override {
         if (address(token) == xcross) {
-            require((msg.value / _exrate) * _exrate == msg.value, BridgeCrossInvalidValueUnit(msg.value));
-            require(msg.value == value + fee, BridgeCrossInsufficientValue(value + fee, msg.value));
-            if (fee > 0) rewardWallet().sendValue(fee);
+            // Handling native token transfers (e.g., ETH, BNB)
+            require((value / _exrate) * _exrate == value, BridgeCrossInvalidValueUnit(msg.value)); // Check for divisibility
+            require(msg.value == value + fee, BridgeCrossInsufficientValue(value + fee, msg.value)); // Verify correct amount received
+            if (fee > 0) rewardWallet().sendValue(fee); // Send fees to the reward wallet
```
Comment (felix-nexus) [inline src/BridgeCross.sol]:
```suggestion
            if (fee != 0) rewardWallet().sendValue(fee); // Send fees to the reward wallet
```

File: src/BridgeCross.sol [hunk 2]
```diff
@@ -17,61 +23,101 @@ contract BridgeCross is BridgeStandard {
...
+     * @param value The amount of tokens being bridged.
+     * @param fee The total fees (gas + service) for the bridge transaction.
+     */
     function _initiateBridge(IERC20 token, address from, uint value, uint fee) internal override {
         if (address(token) == xcross) {
-            require((msg.value / _exrate) * _exrate == msg.value, BridgeCrossInvalidValueUnit(msg.value));
-            require(msg.value == value + fee, BridgeCrossInsufficientValue(value + fee, msg.value));
-            if (fee > 0) rewardWallet().sendValue(fee);
+            // Handling native token transfers (e.g., ETH, BNB)
+            require((value / _exrate) * _exrate == value, BridgeCrossInvalidValueUnit(msg.value)); // Check for divisibility
```
Comment (felix-nexus) [inline src/BridgeCross.sol]:
```suggestion
            require(value % _exrate == 0, BridgeCrossInvalidValueUnit(msg.value)); // Check for divisibility
```

File: src/BridgeCross.sol [hunk 3]
```diff
@@ -17,61 +23,101 @@ contract BridgeCross is BridgeStandard {
...
+            require(msg.value == value + fee, BridgeCrossInsufficientValue(value + fee, msg.value)); // Verify correct amount received
+            if (fee > 0) rewardWallet().sendValue(fee); // Send fees to the reward wallet
         } else {
-            if (fee > 0) token.safeTransferFrom(from, rewardWallet(), fee);
+            // Handling ERC20 token transfers
+            if (fee > 0) token.safeTransferFrom(from, rewardWallet(), fee); // Transfer fees to the reward wallet
```
Comment (felix-nexus) [inline src/BridgeCross.sol]:
```suggestion
            if (fee != 0) token.safeTransferFrom(from, rewardWallet(), fee); // Transfer fees to the reward wallet
```

File: src/BridgeCross.sol [hunk 4]
```diff
@@ -17,61 +23,101 @@ contract BridgeCross is BridgeStandard {
...
+     * @param value The amount of tokens being bridged.
+     * @return ok True if the finalization was successful, false otherwise.
+     * @return reason The reason for failure if the finalization was unsuccessful.
+     */
     function _finalizeBridge(IERC20 token, address to, uint value)
         internal
         override
         returns (bool ok, bytes memory reason)
     {
         if (value > 0) {
```
Comment (felix-nexus) [inline src/BridgeCross.sol]:
```suggestion
        if (value != 0) {
```

File: src/abstract/BridgeStandard.sol [hunk 5]
```diff
@@ -42,54 +52,62 @@ abstract contract BridgeStandard is
...
+        bool permit,
         bytes[] extraData
     );
     event BridgeFinalized(uint indexed index, IERC20 indexed token, address indexed to, uint value, uint time);
     event BridgeFinalizeReverted(uint indexed index);
+    event BridgeFeeCharged(
+        uint indexed index, IERC20 indexed token, address indexed account, uint gasFee, uint serviceFee
+    );
 
     uint internal constant _exrate = 100; // cross : xcross, 1 : 100
```
Comment (felix-nexus) [inline src/abstract/BridgeStandard.sol]: It would be better to represent constants using uppercase camel case.

File: src/BridgeTokenInfo.sol [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    function updateTokenInfoMany(TokenInfo[] memory tokenInfoList) external {
+        for (uint i = 0; i < tokenInfoList.length; i++) {
+            updateTokenInfo(
+                tokenInfoList[i].token,
+                tokenInfoList[i].minimumValue,
+                tokenInfoList[i].gasFee,
+                tokenInfoList[i].serviceFee
+            );
+        }
```
Comment (felix-nexus) [inline src/BridgeTokenInfo.sol]:
```suggestion
        for (uint i = 0; i < tokenInfoList.length;) {
            updateTokenInfo(
                tokenInfoList[i].token,
                tokenInfoList[i].minimumValue,
                tokenInfoList[i].gasFee,
                tokenInfoList[i].serviceFee
            );
            unchecked {
                ++i;
            }
        }
```

File: src/BridgeTokenInfo.sol [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+contract BridgeTokenInfo is Ownable, TokenStorage, IBridgeTokenInfo {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    error BridgeTokenInfoInvalidToken(address token);
+
+    event TokenInfoUpdated(address indexed token, uint minimumValue, uint gasFee, uint serviceFee);
+    event TokenInfoRemoved(address indexed token);
+
+    mapping(address => TokenInfo) private _tokenInfo;
+    uint private _denominator = 1000;
```
Comment (felix-nexus) [inline src/BridgeTokenInfo.sol]: It would be good to use this as a constant.

File: src/abstract/BridgeStandard.sol [hunk 8]
```diff
@@ -42,139 +53,167 @@ abstract contract BridgeStandard is
...
-    uint internal constant _exrate = 100; // cross : xcross, 1 : 100
+    uint internal constant EX_RATE = 100; // cross : xcross, 1 : 100
     bytes32 private constant FINALIZE_TYPEHASH =
-        keccak256("Finalize(uint256 index,address token,address to,uint256 value)");
+        keccak256("Finalize(uint256 index,address token,address to,uint256 value,bytes[] extraData)");
 
-    IBridgeFeeManager public BridgeFeeManager;
-    uint private _initializedAt;
+    IBridgeTokenInfo public bridgeTokenInfo;
+    address internal coin; // Address representing the native coin on the source chain (e.g., XCROSS, ETH, BNB)
```
Comment (felix-nexus) [inline src/abstract/BridgeStandard.sol]: coin may be better used as a constant rather than a variable.

File: src/abstract/BridgeStandard.sol [hunk 9]
```diff
@@ -42,139 +53,167 @@ abstract contract BridgeStandard is
...
 
         require(address(rewardWallet_) != address(0), BridgeStandardCanNotZeroAddress("rewardWallet"));
-        // require(address(_BridgeFeeManager) != address(0), CanNotZeroAddress("BridgeFeeManager")); // allow zero address
+        // require(address(_bridgeTokenInfo) != address(0), CanNotZeroAddress("_bridgeTokenInfo")); // allow zero address
 
-        _initializedAt = block.timestamp;
-        BridgeFeeManager = IBridgeFeeManager(_BridgeFeeManager);
+        coin = address(1);
+        _initializedAt = block.number;
+        bridgeTokenInfo = IBridgeTokenInfo(_bridgeTokenInfo);
```
Comment (felix-nexus) [inline src/abstract/BridgeStandard.sol]:
```suggestion
        if (_bridgeTokenInfo != address(0)) bridgeTokenInfo = IBridgeTokenInfo(_bridgeTokenInfo);
```

File: src/abstract/BridgeStandard.sol [hunk 10]
```diff
@@ -190,21 +229,24 @@ abstract contract BridgeStandard is
...
-     * @notice Retries the finalization of a bridge transfer that was previously reverted
-     * @param index The index of the bridge transfer to retry
+     * @notice Retries a reverted finalize transaction.
+     * @param index The index of the reverted transaction.
+     * @return True if the retry was successful.
      */
-    function retryFinalize(uint index) external whenNotPaused onlyValidator nonReentrant returns (bool) {
+    function retryFinalize(uint index) public whenNotPaused nonReentrant returns (bool) {
         require(_revertedIndex.contains(index), BridgeStandardNotExistingIndex(index));
         FinalizeArguments memory args = _revertedArguments[index];
```
Comment (felix-nexus) [inline src/abstract/BridgeStandard.sol:260]: Is it correct that FinalizeArguments.extraData is not used at this moment?

File: src/abstract/ValidatorManager.sol [hunk 11]
```diff
@@ -78,7 +91,7 @@ abstract contract ValidatorManager is OwnableUpgradeable, EIP712Upgradeable, IVa
```
Comment (felix-nexus) [inline src/abstract/ValidatorManager.sol:102]: It seems that if the _validate function returns false, it is always reverted. It might be better to handle the revert within this function instead of returning a bool.

Comment (felix-nexus) [reply src/abstract/ValidatorManager.sol:102]: It would be possible to handle the revert with more detailed reasons.

File: src/abstract/BridgeStandard.sol [hunk 12]
```diff
@@ -190,21 +229,24 @@ abstract contract BridgeStandard is
...
+     * @param sigs A 2D array of signatures.
+     * @return True if all bridge transactions were finalized successfully.
      */
     function finalizeBatch(FinalizeArguments[] calldata args, bytes[][] memory sigs) external payable returns (bool) {
-        for (uint i = 0; i < args.length; i++) {
+        for (uint i = 0; i < args.length; ++i) {
             finalize(args[i].index, args[i].token, args[i].to, args[i].value, args[i].extraData, sigs[i]);
         }
         return true;
     }
```
Comment (felix-nexus) [inline src/abstract/BridgeStandard.sol]:
```suggestion
    function finalizeBatch(FinalizeArguments[] memory args, bytes[][] memory sigs) external payable returns (bool) {
        for (uint i = 0; i < args.length; ++i) {
            FinalizeArguments memory arg = args[i];
            finalize(arg.index, arg.token, arg.to, arg.value, arg.extraData, sigs[i]);
        }
        return true;
    }
```

Comment (felix-nexus) [reply src/abstract/BridgeStandard.sol]: There are limitations on extradata in finalize.

File: src/abstract/BridgeStandard.sol [hunk 13]
```diff
@@ -219,130 +261,181 @@ abstract contract BridgeStandard is
...
-     * @return service The calculated service fee
+     * @notice Internal function to execute a permit call for ERC20 tokens.
+     * @param permitArgs The permit arguments struct containing necessary data for the permit call.
+     * This includes the token, account, value, deadline, v, r, and s parameters.
+     * @dev Reverts if the permit call fails or if the token is not a valid ERC20Permit contract.
      */
-    function calculateFee(IERC20 token, uint value) external view returns (uint gas, uint service) {
-        (gas, service) = BridgeFeeManager.calculateFee(address(token), value);
-    }
+    function _permitCall(PermitArguments memory permitArgs) private {
```
Comment (felix-nexus) [inline src/abstract/BridgeStandard.sol:393]: "Is this function significantly more efficient than simply calling it like this: token.permit(owner, spender, value, deadline, v, r, s);?"

File: src/BridgeCross.sol [hunk 14]
```diff
@@ -9,79 +9,126 @@ import {Create2} from "@openzeppelin/contracts/utils/Create2.sol";
...
+     */
+    function initialize(address crossMintableERC20Code, address rewardWallet_, address _bridgeTokenInfo)
         external
         initializer
     {
-        __BridgeStandard_init(rewardWallet_, BridgeFeeManager);
-        xcross = address(1);
+        __BridgeStandard_init(rewardWallet_, _bridgeTokenInfo);
         _crossMintableERC20Code = ICrossMintableERC20Code(crossMintableERC20Code);
+        weth = addTokenDeploy(IERC20(address(1)), "ETH", 18);
```
Comment (felix-nexus) [inline src/BridgeCross.sol]:
```suggestion
        weth = addTokenDeploy(IERC20(coin), "ETH", 18);
```

File: src/BridgeEthereum.sol [hunk 15]
```diff
@@ -2,50 +2,116 @@
...
+     * @notice Initializes the BridgeEthereum contract.
+     * @param _cross The address of the wrapped native token on the destination chain (e.g., WETH).
+     * @param rewardWallet_ The address of the reward wallet to receive fees.
+     * @param _bridgeTokenInfo The address of the BridgeTokenInfo contract.
+     */
+    function initialize(IERC20 _cross, address rewardWallet_, address _bridgeTokenInfo) public initializer {
+        require(address(_cross) != address(0), BridgeEthereumCanNotZeroAddress("_cross"));
+        __BridgeStandard_init(rewardWallet_, _bridgeTokenInfo);
+        cross = _cross;
+        addToken(_cross, IERC20(address(1)));
```
Comment (felix-nexus) [inline src/BridgeEthereum.sol]:
```suggestion
        addToken(_cross, IERC20(coin));
```

File: src/BridgeTokenInfo.sol [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+
+    IPriceFeed public priceFeed;
+    address public wcoin; // crosschain wrapped eth
+    uint private constant DENOMINATOR = 1000;
+    uint private _exFee = 10;
+    uint private _gasPrice = 5 gwei;
+    uint private _finalizeGas = 100000;
+
+    // mapping(IERC20 => uint) public exchangeFee;
+    mapping(IERC20 => Token) public _tokenInfo;
```
Comment (felix-nexus) [inline src/BridgeTokenInfo.sol:41]: Are you planning to use it as a private variable?

File: src/BridgeTokenInfo.sol [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+
+    constructor(address _wcoin) Ownable(_msgSender()) {
+        require(_wcoin != address(0), BridgeTokenInfoCanNotZeroAddress("_wcoin"));
+        wcoin = _wcoin;
+    }
+
+    function denominator() external pure returns (uint) {
```
Comment (felix-nexus) [inline src/BridgeTokenInfo.sol:48]: If we change DENOMINATOR to public, this function might be removed.

File: src/PriceFeedCross.sol [hunk 18]
```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.28;
+
+import {TokenStorage} from "./abstract/TokenStorage.sol";
+import {ValidatorManager} from "./abstract/ValidatorManager.sol";
+import {IPriceFeed} from "./interface/IPriceFeed.sol";
+import {IERC20Metadata} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+
+library PriceFeedLib {
```
Comment (felix-nexus) [inline src/PriceFeedCross.sol]: It would be better to place the library in a new lib directory.

File: src/PriceFeedCross.sol [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+    error PriceFeedCanNotZeroAddress(string name);
+    error PriceFeedCanNotZeroValue(string name);
+    error PriceFeedNotExistToken(address token);
+    error PriceFeedInvalidLength();
+    error PriceFeedInvalidPriceAt(uint at, uint blocktime);
+
+    event PriceFeedPriceUpdated(address indexed token, uint price, uint timestamp);
+    event PriceFeedDeadlineChanged(uint deadline);
+
+    address public coin;
```
Comment (felix-nexus) [inline src/PriceFeedCross.sol]: coin may be better used as a constant rather than a variable.

File: src/PriceFeedCross.sol [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+                ++i;
+            }
+        }
+    }
+
+    /// @notice Retrieves the price data for a single token.
+    /// @param token The address of the token.
+    /// @return data A PriceData struct containing the price information for the token.
+    function getPrice(address token) public view returns (PriceData memory data) {
+        require(contains(token), PriceFeedNotExistToken(token));
```
Comment (felix-nexus) [inline src/PriceFeedCross.sol:69]: "If you revert here, it seems that it will also affect the getPrices function above. Is that what you intended?"

File: src/PriceFeedCross.sol [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+        // (1 xcross == 0.001 dollar)
+        prices[coin] = PriceData(coin, (10 ** dollarDecimals) / 1000, type(uint).max);
+    }
+
+    /// @notice Retrieves the valid price for a given token.
+    /// @param token The address of the token.
+    /// @return price The price of the token.
+    /// @return isValid True if the price is valid, false otherwise.
+    function getValidPrice(address token) external view returns (uint price, bool isValid) {
+        require(contains(token), PriceFeedNotExistToken(token));
```
Comment (felix-nexus) [inline src/PriceFeedCross.sol]: I think it would be better to return false for isValid instead of reverting.

File: src/PriceFeedCross.sol [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+    error PriceFeedNotExistToken(address token);
+    error PriceFeedInvalidLength();
+    error PriceFeedInvalidPriceAt(uint at, uint blocktime);
+
+    event PriceFeedPriceUpdated(address indexed token, uint price, uint timestamp);
+    event PriceFeedDeadlineChanged(uint deadline);
+
+    address public coin;
+    uint public deadline;
+    uint public dollarDecimals;
```
Comment (felix-nexus) [inline src/PriceFeedCross.sol]: It would be better to use dollarDecimals as a constant if it's not a value that gets modified.

File: src/abstract/BridgeStandard.sol (hunk 10 above)
Comment (zayden-nexus) [reply src/abstract/BridgeStandard.sol:260]: The ExtraData is reserved for processing when needed.

File: src/PriceFeedCross.sol (hunk 20 above)
Comment (zayden-nexus) [reply src/PriceFeedCross.sol:69]: I have modified the getPrices request to return a zero value if it contains an invalid token.
Thank you.

**Review by felix-nexus — APPROVED** [review]
LGTM.

**Comment by zayden-nexus** [comment]
This pull request is being closed as it involves significant feature changes, requiring a new pull request to be submitted.
